crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,450 @@
+    fn new(
+        consensus_adapter: Arc<dyn SubmitToConsensus>,
+        epoch_store: Weak<AuthorityPerEpochStore>,
+        epoch: EpochId,
+        session_id: ObjectID,
+        sender_address: PeraAddress,
+        number_of_parties: usize,
+    ) -> Self {
+        Self {
+            status: MPCSessionStatus::Active,
+            pending_messages: HashMap::new(),
+            consensus_adapter: consensus_adapter.clone(),
+            epoch_store: epoch_store.clone(),
+            epoch_id: epoch,
+            session_id,
+            sender_address,
+            party: None,
+            number_of_parties,
+        }
+    }
+
+    fn epoch_store(&self) -> PeraResult<Arc<AuthorityPerEpochStore>> {
+        self.epoch_store
+            .upgrade()
+            .ok_or(PeraError::EpochEnded(self.epoch_id))
+    }
+
+    /// Advances the MPC instance and optionally return a message the validator wants to send to the other MPC parties.
+    /// Uses the existing party if it exists, otherwise creates a new one, as this is the first advance.
+    fn advance(&mut self, auxiliary_input: &P::AuxiliaryInput) -> PeraResult {
+        let optional_party = mem::take(&mut self.party);
+
+        /// Gets the instance existing party or creates a new one if this is the first advance
+        let party: P = if let Some(existing_party) = optional_party {
+            existing_party
+        } else {
+            let mut parties = HashSet::new();
+            for i in 0..self.number_of_parties {
+                parties.insert(i as PartyID);
+            }
+            P::new(
+                parties,
+                authority_name_to_party_id(self.epoch_store()?.name, &*(self.epoch_store()?))?,
+            )
+        };
+        let Ok(advance_result) =
+            party.advance(self.pending_messages.clone(), auxiliary_input, &mut OsRng)
+        else {
+            // TODO (#263): Mark and punish the malicious validators that caused this advance to fail
+            self.pending_messages.clear();
+            return Ok(());
+        };
+        let msg = match advance_result {
+            AdvanceResult::Advance((message, party)) => {
+                self.pending_messages.clear();
+                self.party = Some(party);
+                self.new_signature_mpc_message(message)
+            }
+            AdvanceResult::Finalize(output) => {
+                // TODO (#238): Verify the output and write it to the chain
+                self.status = MPCSessionStatus::Finalizing(output.clone().into());
+                self.new_proof_mpc_output_message(output.into())
+            }
+        };
+
+        let consensus_adapter = Arc::clone(&self.consensus_adapter);
+        let epoch_store = Arc::clone(&self.epoch_store()?);
+        if let Some(msg) = msg {
+            /// Spawns sending this message asynchronously the [`self.advance`] function will stay synchronous
+            /// and can be parallelized with Rayon.
```

Comment:
// Spawn sends this message asynchronously via a thread
            // so that [`self.advance`] function will stay synchronous,
            // and can be parallelized with Rayon.